### PR TITLE
Rollout Support StatefulSet & Advanced StatefulSet

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: veophi [vec.g.sun@gmail.com](mailto:vec.g.sun@gmail.com)

### Ⅰ. Describe what this PR does

- [x] Support Advanced StatefulSet
- [x] Support Native StatefulSet
- [x] Unit Test & E2E Test

### II. Related Issue
#28

### CODE REVIEW

File: pkg/controller/batchrelease/workloads/statefulset_like_controller.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+func (c *StatefulSetLikeController) IsBatchReady(canaryReplicasGoal, stableReplicasGoal int32) (bool, error) {
+	workloadInfo, err := c.GetWorkloadInfo()
+	if err != nil {
+		return false, err
+	}
+
+	// make sure all pod is available, and the canary goal is met
+	ready := workloadInfo.Status.ReadyReplicas == workloadInfo.Status.Replicas &&
```
Comment (furykerry) [inline pkg/controller/batchrelease/workloads/statefulset_like_controller.go:163]: condition workloadInfo.Status.ReadyReplicas == workloadInfo.Status.Replicas is too strict, there is maybe cases where old replicas are not ready

File: pkg/controller/batchrelease/workloads/unified_workload_control_plane.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	}
+
+	// If the workload has been promoted, no need to go on
+	if workloadInfo.Status.UpdatedReplicas == *workloadInfo.Replicas {
+		err = fmt.Errorf("%v update revision has been promoted, no need to reconcile", workloadInfo.GVKWithName)
+		return false, err
+	}
+
+	// If the workload is not paused and is not under our control
+	if !workloadInfo.Paused {
```
Comment (furykerry) [inline pkg/controller/batchrelease/workloads/unified_workload_control_plane.go:108]: we shall also unpause statefulset in resumeStableWorkload e.g. set partition to nil

File: pkg/controller/batchrelease/workloads/statefulset_like_controller.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+func (c *StatefulSetLikeController) GetWorkloadInfo() (*util.WorkloadInfo, error) {
+	set, err := c.GetWorkloadObject()
+	if err != nil {
+		return nil, err
+	}
+
+	workloadInfo := util.ParseWorkloadInfo(set, c.namespacedName)
+	if util.ParseInt32PartitionFrom(set) == *workloadInfo.Replicas {
```
Comment (furykerry) [inline pkg/controller/batchrelease/workloads/statefulset_like_controller.go:74]: util.ParseInt32PartitionFrom(set) >= *workloadInfo.Replicas

File: pkg/util/controller_finder.go [hunk 4]
```diff
@@ -208,6 +210,46 @@ func (r *ControllerFinder) getDeployment(namespace string, ref *rolloutv1alpha1.
...
 
+func (r *ControllerFinder) getStatefulSetLikeWorkload(namespace string, ref *rolloutv1alpha1.WorkloadRef) (*Workload, error) {
+	if ref == nil || ref.Kind != AStatefulSetGVK.Kind {
+		return nil, nil
+	}
+
+	unifiedObject := &unstructured.Unstructured{}
+	unifiedObjectKey := types.NamespacedName{Name: ref.Name, Namespace: namespace}
+	unifiedObject.SetGroupVersionKind(schema.FromAPIVersionAndKind(ref.APIVersion, ref.Kind))
+	err := r.Get(context.TODO(), unifiedObjectKey, unifiedObject)
```
Comment (FillZpp) [inline pkg/util/controller_finder.go:221]: We'd better not get unstructured object here.

**Pull request closed without merging**